architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,495 @@
+- Remove legacy code path from UCP routing logic. Now only registered types are supported.
+
+Dependency: Stage 1
+
+Exit criteria: all functional tests passing using new routing behavior in UCP. All built-in resource types can be queried with the CLI.
+
+#### Stage 3: CLI enhancements
+
+- Remove hardcoded resource type validation from the CLI and replace with queries of the resource provider/type APIs.
+- Add support for `rad resource create`. A new CLI command that can create resources of any type. This is needed to properly exercise the user-defined types feature until Bicep support comes online.
+
+Dependency: Stage 1
+
+Exit criteria: existing tests passing. Add functional test for `rad resource create` using an existing built-in resource.
+
+#### Stage 4: Dynamic Resource Provider (Dynamic-RP)
+
+- Implement dynamic resource provider with support for CRUDL operations.
+  - No support for validation (yet).
+  - The dynamic resource provider requires the use of Recipes.
+- Implement integration tests and functional tests for Dynamic-RP's basic resource lifecycle using `rad resource create`.
+
+Dependency: Stage 2
+
+Exit criteria: new functional tests exercise the dynamic resource provider.
+
+#### Stage 5: Bicep Support
+
+- Implement a TypeSpec generator that can generate a resource provider manifest and Bicep extension manifest.
+- (If necessary): Update deployment engine logic to generate correct resource ids/URLs for user-defined types.
+- Implement functional tests that exercise user-defined types with Bicep.
+
+Dependency: Stage 4 needed for functional testing, no dependency to begin work.
+
+Exit criteria: new functional tests that exercise the dynamic resource provider with Bicep. 
+
+### Phase 2: Completing the UDT feature
+
+Note: at this point the UDT feature can do useful work end-to-end. The stages in this section make UDT feature-complete.
+
+#### Stage 6: Documentation and Samples
+
+We can begin adding documentation and samples once we reach **Stage 5**. Starting as soon as the features are useful allows us to incremental improve the documentation as more features come online.
+
+Dependency: Stage 5
+
+Exit criteria:
+
+- Tutorial created for an end-to-end of authoring a user-defined type and using it.
+- Conceptual documentation for extensibility.
+- Call-to-action for users to try it out and give feedback.
+
+#### Stage 7: Schema Validation
+
+- Accept an OpenAPI v3 schema as part of a resource type.
+- Implement OpenAPI validation in the dynamic resource provider.
+
+Dependency: Stage 4
+
+Exit criteria: new negative tests for the dynamic resource provider.
+
+#### Stage 8: Creation of community repo and contribution process
+
+At this stage we don't expect breaking changes going forward. UDTs will have reasonable guardrails in place (schema-based validation) so they are usable for many cases. 
+
+- Create a community *contrib* repository for user-defined types.
```

Comment:
How do we make sure that the new user-defined type is working fine and as expected? Are we going to require contributors to write functional tests for the new user-defined type? Because I believe that could be the way to make sure the new type is working.

Reply:
> Are we going to require contributors to write functional tests for the new user-defined type?

I hope not, that sounds like a lot of work. One of the great things about declarative functionality is that less testing is required. If you didn't write the code, you don't have anything to test.

I think we **do** need functional tests for community supported recipes.